### PR TITLE
[RORDEV-431] Custom audit cluster

### PR DESCRIPTION
**Pull request by pondzix**
🚀New (ES) Configuration for custom audit cluster
🧐Enhancement (ES) Separate "audit" section for all audit settings

### CODE REVIEW

File: core/src/test/scala/tech/beshu/ror/unit/acl/factory/AuditingSettingsTests.scala [hunk 1]
```diff
@@ -209,8 +214,40 @@ class AuditingSettingsTests extends AnyWordSpec with Inside {
...
           auditingSettings.logSerializer shouldBe a[DeprecatedAuditLogSerializerAdapter[_]]
+          auditingSettings.customAuditCluster shouldBe None
+        }
+      }
+      "custom audit cluster is set" in {
+        val config = rorConfigFromUnsafe(
+          """
+            |readonlyrest:
+            |  audit_collector: true
+            |  audit_cluster: ["1.1.1.1"]
```
Comment (coutoPL) [inline core/src/test/scala/tech/beshu/ror/unit/acl/factory/AuditingSettingsTests.scala:225]: missing test with empty list

File: core/src/test/scala/tech/beshu/ror/unit/acl/logging/AuditingToolTests.scala [hunk 2]
```diff
@@ -49,7 +49,8 @@ class AuditingToolTests extends AnyWordSpec with MockFactory {
         val auditingTool = new AuditingTool(
           AuditingTool.Settings(
             RorAuditIndexTemplate.from("'test_'yyyy-MM-dd").right.get,
-            new DefaultAuditLogSerializer
+            new DefaultAuditLogSerializer,
+            None
```
Comment (coutoPL) [inline core/src/test/scala/tech/beshu/ror/unit/acl/logging/AuditingToolTests.scala:53]: are we able to test remote cluster here?

File: es714x/src/main/scala/tech/beshu/ror/es/IndexLevelActionFilter.scala [hunk 3]
```diff
@@ -158,6 +171,33 @@ class IndexLevelActionFilter(clusterService: ClusterService,
...
+    val hosts = auditCluster.uris.map { uri =>
+      new HttpHost(uri.host, uri.port.getOrElse(9200), uri.scheme)
+    }.toList
+
+    new RestHighLevelClient(
+      RestClient
+        .builder(hosts: _*)
+        .setHttpClientConfigCallback(
+          (httpClientBuilder: HttpAsyncClientBuilder) => {
+            //TODO What about cert verification?
```
Comment (coutoPL) [inline es714x/src/main/scala/tech/beshu/ror/es/IndexLevelActionFilter.scala:185]: It can be done line that atm

File: es714x/src/main/scala/tech/beshu/ror/es/IndexLevelActionFilter.scala [hunk 4]
```diff
@@ -158,6 +171,33 @@ class IndexLevelActionFilter(clusterService: ClusterService,
         rorInstanceState.set(RorInstanceStartingState.NotStarted(StartingFailureException.from(startingFailureException)))
     }
   }
+
+  private def createEsHighLevelClient(auditCluster: AuditCluster) = {
```
Comment (coutoPL) [inline es714x/src/main/scala/tech/beshu/ror/es/IndexLevelActionFilter.scala:175]: why this is not a part of HighLevelClientAuditSinkService?

**Comment by sscarduzio** [comment]
Hey @pondzix! Good to see you, mate! 😃

File: core/src/main/scala/tech/beshu/ror/accesscontrol/factory/GlobalSettings.scala [hunk 5]
```diff
@@ -16,14 +16,13 @@
...
 
-import tech.beshu.ror.accesscontrol.domain.{RorAuditIndexTemplate, RorConfigurationIndex}
+import tech.beshu.ror.accesscontrol.domain.RorConfigurationIndex
 
 final case class GlobalSettings(showBasicAuthPrompt: Boolean,
                                 forbiddenRequestMessage: String,
                                 flsEngine: GlobalSettings.FlsEngine,
                                 configurationIndex: RorConfigurationIndex,
-                                usernameCaseMapping: GlobalSettings.UsernameCaseMapping,
-                                indexAuditTemplate: Option[RorAuditIndexTemplate])
```
Comment (pondzix) [inline core/src/main/scala/tech/beshu/ror/accesscontrol/factory/GlobalSettings.scala:26]: Do we need `indexAuditTemplate` in global settings? It looks like it's unused, so I deleted it.

File: core/src/test/scala/tech/beshu/ror/unit/acl/logging/AuditingToolTests.scala (hunk 2 above)
Comment (pondzix) [reply core/src/test/scala/tech/beshu/ror/unit/acl/logging/AuditingToolTests.scala]: Nope, we provide `mock[AuditSinkService]` to `AuditingTool`. Using audit cluster settings here is confusing as it's not used by `AuditingTool` at all, but at the layer above in `ReadonlyRest`. It's here, because we have to pass something to provide full audit settings. Maybe it means `AuditingTool.Settings` should be restructured, so we don't have to use cluster settings here...? Anyway, we have it covered in integration tests ;)

**Review by coutoPL — CHANGES_REQUESTED** [review]

File: integration-tests/src/test/scala/tech/beshu/ror/integration/suites/RemoteClusterAuditingToolsSuite.scala [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+package tech.beshu.ror.integration.suites
+
+import cats.data.NonEmptyList
+import tech.beshu.ror.integration.suites.base.BaseAuditingToolsSuite
+import tech.beshu.ror.integration.suites.base.support.{BaseManyEsClustersIntegrationTest, MultipleClientsSupport}
+import tech.beshu.ror.utils.containers.providers.ClientProvider
+import tech.beshu.ror.utils.containers.{EsClusterContainer, EsClusterSettings, EsContainer, EsContainerCreator}
+import tech.beshu.ror.utils.elasticsearch.ElasticsearchTweetsInitializer
+
+trait RemoteClusterAuditingToolsSuite
```
[the remark this answers is not in the document]
Comment (pondzix) [reply integration-tests/src/test/scala/tech/beshu/ror/integration/suites/RemoteClusterAuditingToolsSuite.scala:27]: I added support for running dependencies for singleton ES, so in this case source ES is shared singleton container and audit node is ES container (without plugin) run as dependency of source cluster. I also adjusted other suites with dependencies to this approach.

File: es55x/src/main/scala/tech/beshu/ror/es/IndexLevelActionFilter.scala [hunk 7]
```diff
@@ -150,8 +159,8 @@ class IndexLevelActionFilter(settings: Settings,
   }
 
   private def startRorInstance() = doPrivileged {
-    new Ror(RorMode.Plugin)
-      .start(env.configFile, new EsAuditSinkService(client), new EsIndexJsonContentService(client))
+    new Ror(RorMode.Plugin, auditSinkCreator)
```
Comment (coutoPL) [inline es55x/src/main/scala/tech/beshu/ror/es/IndexLevelActionFilter.scala:162]: don't you think we have an inconsistency here? We pass IndexJsonContentService implementation in the start method but AuditSinkCreator in the constructor. Can we pick one approach?

File: es66x/src/main/scala/tech/beshu/ror/es/services/HighLevelClientAuditSinkService.scala [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+import java.security.cert.X509Certificate
+import javax.net.ssl.{SSLContext, TrustManager, X509TrustManager}
+
+class HighLevelClientAuditSinkService(client: RestHighLevelClient)
+                                     (implicit scheduler: Scheduler)
+  extends AuditSinkService
+    with Logging {
+
+  override def submit(indexName: String, documentId: String, jsonRecord: String): Unit = {
```
Comment (coutoPL) [inline es66x/src/main/scala/tech/beshu/ror/es/services/HighLevelClientAuditSinkService.scala:39]: oh, can we add the implicit `RequestContext.Id` parameter to the method? We could use it in logs

File: plugin-metadata/plugin-security.policy [hunk 9]
```diff
@@ -6,4 +6,5 @@ grant {
   permission java.lang.reflect.ReflectPermission "suppressAccessChecks";
   permission java.util.PropertyPermission "*", "read,write";
   permission java.net.SocketPermission "*", "accept, resolve, connect";
+  permission java.net.NetPermission "getProxySelector";
```
Comment (coutoPL) [inline plugin-metadata/plugin-security.policy:9]: why do we need it?

**Review by coutoPL — CHANGES_REQUESTED** [review]

File: plugin-metadata/plugin-security.policy (hunk 9 above)
Comment (pondzix) [reply plugin-metadata/plugin-security.policy:9]: es60x needs this one to create HLC

File: es66x/src/main/scala/tech/beshu/ror/es/services/HighLevelClientAuditSinkService.scala (hunk 8 above)
Comment (pondzix) [reply es66x/src/main/scala/tech/beshu/ror/es/services/HighLevelClientAuditSinkService.scala:39]: It seems request context id is passed as `documentId` and it's already logged.

File: es55x/src/main/scala/tech/beshu/ror/es/IndexLevelActionFilter.scala (hunk 7 above)
Comment (pondzix) [reply es55x/src/main/scala/tech/beshu/ror/es/IndexLevelActionFilter.scala]: Your suggestion encouraged me to do some changes:

* convert `ReadonlyRest` trait to class, no more `Ror` class
* remove all params from `start` method, pass everything in constructor of `ReadonlyRest`
* extract `RorInstance` to separate file
* extract ssl tests from `ReadonlyRestStartingTests` to `SslConfigurationTest`

WDYT?

File: es60x/src/main/scala/tech/beshu/ror/es/services/HighLevelClientAuditSinkService.scala [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        case Right(resp) if resp.status().getStatus / 100 == 2 =>
+        case Right(resp) =>
+          logger.error(s"Cannot submit audit event [index: $indexName, doc: $documentId] - response code: ${resp.status().getStatus}")
+        case Left(ex) =>
+          logger.error(s"Cannot submit audit event [index: $indexName, doc: $documentId]", ex)
+      }
+  }
+
+  override def close(): Unit = {
+    client.close()
```
Comment (coutoPL) [inline es60x/src/main/scala/tech/beshu/ror/es/services/HighLevelClientAuditSinkService.scala:56]: TBH the one who creates the client should close it. But it'd require to change too much (eg. create method could return cat's Resource). So, let's leave it as it's now, but let's make the HighLevelClientAuditSinkService's constructor private

File: es55x/src/main/scala/tech/beshu/ror/es/IndexLevelActionFilter.scala (hunk 7 above)
Comment (coutoPL) [reply es55x/src/main/scala/tech/beshu/ror/es/IndexLevelActionFilter.scala]: it's much better now.

File: plugin-metadata/plugin-security.policy (hunk 9 above)
Comment (coutoPL) [reply plugin-metadata/plugin-security.policy:9]: oh, ok. So, I'd like you to do one more improvement - please let's:
* copy the `plugin-metadata` folder to each es module.
* leave the permission only for es60x
* remove `plugin-metadata` from root folder
* adapt `resolvePluginDescriptorTemplate` task

**Comment by coutoPL** [comment]
These are the last change requests. Please remember about documentation and changelog-style description of the PR.

I see, that thanks to your changes, we gained a big improvement in tests execution time. Great job!!!

**Comment by pondzix** [comment]
Docs: https://github.com/beshu-tech/readonlyrest-docs/pull/105

**Review by coutoPL — APPROVED** [review]
thanks :)
